File: 2267-check-if-there-is-a-valid-parentheses-string-path/2267-check-if-there-is-a-valid-parentheses-string-path.go
func dfs(grid [][]byte, dp [][][]int, y, x, st int) bool {
	if st < 0 {
		return false
	}
	if y < 0 || x < 0 {
		return false
	}
	if dp[y][x][st] != -1 {
		return dp[y][x][st] == 1
	}
	tmp := st
	if grid[y][x] == ')' {
		st++
	} else {
		st--
	}
	if x == 0 && y == 0 {
		return st == 0
	}
	result := (dfs(grid, dp, y-1, x, st) || dfs(grid, dp, y, x-1, st))
	if result {
		dp[y][x][tmp] = 1
		return true
	} else {
		dp[y][x][tmp] = 0
		return false
	}
}

func hasValidPath(grid [][]byte) bool {
    Y := len(grid)
    X := len(grid[0])
    maxState := Y + X
    dp := make([][][]int, Y)
    for i := range dp {
        dp[i] = make([][]int, X)
        for j := range dp[i] {
            dp[i][j] = make([]int, maxState+1)
            for k := range dp[i][j] {
                dp[i][j][k] = -1
            }
        }
    }
    return dfs(grid, dp, Y-1, X-1, 0) 
}

File: 2267-check-if-there-is-a-valid-parentheses-string-path/2267-check-if-there-is-a-valid-parentheses-string-path.cpp
class Solution {
public:
    vector<vector<vector<int>>> dp;
    int m, n;
    bool dfs(vector<vector<char>>& grid, int i, int j, int st) {
        if (i < 0 || j < 0)
            return false;
        if (dp[i][j][st] != -1)
            return dp[i][j][st];
        int tmp=st;
        if (grid[i][j] == ')')
            st++;
        else
            st--;
        if (j == 0 && i == 0)
            return !st;
        if (st < 0)
            return false;
        return dp[i][j][tmp] =
                   (dfs(grid, i - 1, j, st) or dfs(grid, i, j - 1, st));
    }
    bool hasValidPath(vector<vector<char>>& grid) {
        m = grid.size(), n = grid[0].size();
        dp.resize(m, vector<vector<int>>(n, vector<int>(m + n, -1)));
        return dfs(grid, m - 1, n - 1, 0);
    }
};